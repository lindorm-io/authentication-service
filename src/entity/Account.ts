import { AccountEvent } from "../enum";
import { EntityBase, EntityCreationError, IEntity, IEntityBaseOptions } from "@lindorm-io/core";
import { Permission } from "@lindorm-io/jwt";

export interface IAccountOTP {
  signature: string;
  uri: string;
}

export interface IAccountPassword {
  signature: string;
  updated: Date;
}

export interface IAccount extends IEntity {
  email: string;
  identityLinked: boolean;
  otp: IAccountOTP;
  password: IAccountPassword;
  permission: string;
}

export interface IAccountOptions extends IEntityBaseOptions {
  email: string;
  identityLinked?: boolean;
  otp?: IAccountOTP;
  password?: IAccountPassword;
  permission?: string;
}

export class Account extends EntityBase implements IAccount {
  private _email: string;
  private _identityLinked: boolean;
  private _otp: IAccountOTP;
  private _password: IAccountPassword;
  private _permission: string;

  constructor(options: IAccountOptions) {
    super(options);
    this._email = options.email.toLowerCase();
    this._identityLinked = options.identityLinked || false;
    this._otp = {
      signature: options.otp?.signature || null,
      uri: options.otp?.uri || null,
    };
    this._password = {
      signature: options.password?.signature || null,
      updated: options.password?.updated || null,
    };
    this._permission = options.permission || Permission.USER;
  }

  public get email(): string {
    return this._email;
  }
  public set email(email: string) {
    this._email = email.toLowerCase();
    this.addEvent(AccountEvent.EMAIL_CHANGED, { email: this._email });
  }

  public get identityLinked(): boolean {
    return this._identityLinked;
  }
  public set identityLinked(identityLinked: boolean) {
    this._identityLinked = identityLinked;
    this.addEvent(AccountEvent.EMAIL_CHANGED, { identityLinked: this._identityLinked });
  }

  public get otp(): IAccountOTP {
    return this._otp;
  }
  public set otp(otp: IAccountOTP) {
    this._otp = otp;
    this.addEvent(AccountEvent.OTP_CHANGED, { otp: this._otp });
  }

  public get password(): IAccountPassword {
    return this._password;
  }
  public set password(password: IAccountPassword) {
    this._password = password;
    this.addEvent(AccountEvent.PASSWORD_CHANGED, { password: this._password });
  }

  public get permission(): string {
    return this._permission;
  }
  public set permission(permission: string) {
    this._permission = permission;
    this.addEvent(AccountEvent.PERMISSION_CHANGED, { permission: this._permission });
  }

  public create(): void {
    for (const evt of this._events) {
      if (evt.name !== AccountEvent.CREATED) continue;
      throw new EntityCreationError("Account");
    }
    this.addEvent(AccountEvent.CREATED, {
      email: this._email,
      permission: this._permission,
      password: this._password,
      otp: this._otp,
      created: this._created,
      updated: this._updated,
    });
  }
}
