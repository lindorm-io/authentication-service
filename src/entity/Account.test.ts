import MockDate from "mockdate";
import { Account } from "./Account";

jest.mock("uuid", () => ({
  v4: jest.fn(() => "be3a62d1-24a0-401c-96dd-3aff95356811"),
}));

MockDate.set("2020-01-01 08:00:00.000");
const date = new Date("2020-01-01 08:00:00.000");

describe("Account.ts", () => {
  let account: Account;

  beforeEach(() => {
    account = new Account({
      email: "email@lindorm.io",
      identityLinked: false,
      otp: { signature: "signature", uri: "uri" },
      password: { signature: "password", updated: date },
      permission: "permission",
    });
  });

  test("should have all data", () => {
    expect(account).toMatchSnapshot();
  });

  test("should have optional data", () => {
    account = new Account({
      email: "email@lindorm.io",
    });

    expect(account).toMatchSnapshot();
  });

  test("should create", () => {
    account.create();
    expect(account.events).toMatchSnapshot();
  });

  test("should always cast email as lowercase", () => {
    account = new Account({
      email: "EMAIL@LINDORM.IO",
    });
    expect(account).toMatchSnapshot();
  });

  test("should get/set email as lowercase", () => {
    expect(account.email).toBe("email@lindorm.io");

    account.email = "NEW-email@lindorm.io";

    expect(account.email).toBe("new-email@lindorm.io");
    expect(account.events).toMatchSnapshot();
  });

  test("should get/set identityLinked", () => {
    expect(account.identityLinked).toBe(false);

    account.identityLinked = true;

    expect(account.identityLinked).toBe(true);
    expect(account.events).toMatchSnapshot();
  });

  test("should get/set otp", () => {
    expect(account.otp).toStrictEqual({ signature: "signature", uri: "uri" });

    account.otp = { signature: "new-signature", uri: "new-uri" };

    expect(account.otp).toStrictEqual({ signature: "new-signature", uri: "new-uri" });
    expect(account.events).toMatchSnapshot();
  });

  test("should get/set password", () => {
    expect(account.password).toStrictEqual({ signature: "password", updated: date });

    account.password = { signature: "new-password", updated: date };

    expect(account.password).toStrictEqual({ signature: "new-password", updated: date });
    expect(account.events).toMatchSnapshot();
  });

  test("should get/set permission", () => {
    expect(account.permission).toBe("permission");

    account.permission = "new-permission";

    expect(account.permission).toBe("new-permission");
    expect(account.events).toMatchSnapshot();
  });
});
