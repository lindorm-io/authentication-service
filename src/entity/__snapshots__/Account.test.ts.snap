// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Account.ts should create 1`] = `
Array [
  Object {
    "date": 2020-01-01T07:00:00.000Z,
    "name": "account_created",
    "payload": Object {
      "created": 2020-01-01T07:00:00.000Z,
      "email": "email@lindorm.io",
      "otp": Object {
        "signature": "signature",
        "uri": "uri",
      },
      "password": Object {
        "signature": "password",
        "updated": 2020-01-01T07:00:00.000Z,
      },
      "permission": "permission",
      "updated": 2020-01-01T07:00:00.000Z,
    },
  },
]
`;

exports[`Account.ts should get/set email 1`] = `
Array [
  Object {
    "date": 2020-01-01T07:00:00.000Z,
    "name": "account_email_changed",
    "payload": Object {
      "email": "new-email",
    },
  },
]
`;

exports[`Account.ts should get/set identityId 1`] = `
Array [
  Object {
    "date": 2020-01-01T07:00:00.000Z,
    "name": "account_email_changed",
    "payload": Object {
      "identityId": "ca02be54-0b11-4d45-b918-f637c3162517",
    },
  },
]
`;

exports[`Account.ts should get/set otp 1`] = `
Array [
  Object {
    "date": 2020-01-01T07:00:00.000Z,
    "name": "account_otp_changed",
    "payload": Object {
      "otp": Object {
        "signature": "new-signature",
        "uri": "new-uri",
      },
    },
  },
]
`;

exports[`Account.ts should get/set password 1`] = `
Array [
  Object {
    "date": 2020-01-01T07:00:00.000Z,
    "name": "account_password_changed",
    "payload": Object {
      "password": Object {
        "signature": "new-password",
        "updated": 2020-01-01T07:00:00.000Z,
      },
    },
  },
]
`;

exports[`Account.ts should get/set permission 1`] = `
Array [
  Object {
    "date": 2020-01-01T07:00:00.000Z,
    "name": "account_permission_changed",
    "payload": Object {
      "permission": "new-permission",
    },
  },
]
`;

exports[`Account.ts should have all data 1`] = `
Account {
  "_created": 2020-01-01T07:00:00.000Z,
  "_email": "email@lindorm.io",
  "_events": Array [],
  "_id": "be3a62d1-24a0-401c-96dd-3aff95356811",
  "_identityId": "1148b4d6-214b-4bc3-9611-feecc065bf3c",
  "_otp": Object {
    "signature": "signature",
    "uri": "uri",
  },
  "_password": Object {
    "signature": "password",
    "updated": 2020-01-01T07:00:00.000Z,
  },
  "_permission": "permission",
  "_updated": 2020-01-01T07:00:00.000Z,
  "_version": 0,
}
`;

exports[`Account.ts should have optional data 1`] = `
Account {
  "_created": 2020-01-01T07:00:00.000Z,
  "_email": "email@lindorm.io",
  "_events": Array [],
  "_id": "be3a62d1-24a0-401c-96dd-3aff95356811",
  "_identityId": null,
  "_otp": Object {
    "signature": null,
    "uri": null,
  },
  "_password": Object {
    "signature": null,
    "updated": null,
  },
  "_permission": "user",
  "_updated": 2020-01-01T07:00:00.000Z,
  "_version": 0,
}
`;
